### ISSUE TITLE
Not able to complete docker_container task anymore

### ISSUE THREAD

**Issue by zevarito** [ansible/ansible#42274]
##### SUMMARY

`docker_container` task doesn't create/update container. It was working fine until upgrade osx to high sierra.

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
```
docker_container
```

##### ANSIBLE VERSION

```
ansible 2.5.5 (detached HEAD 5f6e1a70a6) last updated 2018/07/03 12:20:17 (GMT -300)
  config file = /Users/zevarito/playbooks/ansible.cfg
  configured module search path = [u'/Users/zevarito/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/zevarito/ansible/lib/ansible
  executable location = /Users/zevarito/ansible/bin/ansible
  python version = 2.7.10 (default, Oct  6 2017, 22:29:07) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]
```

##### CONFIGURATION
```
ANSIBLE_PIPELINING(/Users/zevarito/playbooks/ansible.cfg) = True
DEFAULT_FORCE_HANDLERS(/Users/zevarito/playbooks/ansible.cfg) = True
DEFAULT_FORKS(/Users/zevarito/playbooks/ansible.cfg) = 50
DEFAULT_HOST_LIST(/Users/zevarito/playbooks/ansible.cfg) = [u'/Users/zevarito/playbooks/inventory/ec2.py']
DEFAULT_REMOTE_USER(/Users/zevarito/playbooks/ansible.cfg) = ubuntu
DEFAULT_TIMEOUT(/Users/zevarito/playbooks/ansible.cfg) = 10
DISPLAY_SKIPPED_HOSTS(/Users/zevarito/playbooks/ansible.cfg) = False
HOST_KEY_CHECKING(/Users/zevarito/playbooks/ansible.cfg) = False
```

##### OS / ENVIRONMENT

`OSX High Sierra`

##### STEPS TO REPRODUCE

```yaml
- name: run API container
  docker_container:
    name: api
    image: "{{ docker_image }}"
    restart: True
    restart_policy: always
    recreate: True
    state: started
    pull: True
    expose:
      - "{{ api_private_port }}"
    ports:
      - "{{ api_private_port }}:{{ api_private_port }}"
    env: "{{ api_env_variables }}"
```

##### EXPECTED RESULTS

I expect the task to be completed fine, it was working for a long time until now, the only thing I can think that have changed on my env is the upgrade of osx high sierra.

##### ACTUAL RESULTS

```
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_ADXrQF/ansible_module_docker_container.py", line 1963, in <module>
    main()
  File "/tmp/ansible_ADXrQF/ansible_module_docker_container.py", line 1956, in main
    cm = ContainerManager(client)
  File "/tmp/ansible_ADXrQF/ansible_module_docker_container.py", line 1609, in __init__
    self.present(state)
  File "/tmp/ansible_ADXrQF/ansible_module_docker_container.py", line 1634, in present
    different, differences = container.has_different_configuration(image)
  File "/tmp/ansible_ADXrQF/ansible_module_docker_container.py", line 1169, in has_different_configuration
    self.parameters.expected_exposed = self._get_expected_exposed(image)
  File "/tmp/ansible_ADXrQF/ansible_module_docker_container.py", line 1555, in _get_expected_exposed
    image_ports = [re.sub(r'/.+$', '', p) for p in (image['ContainerConfig'].get('ExposedPorts') or {}).keys()]
TypeError: tuple indices must be integers, not str

fatal: [52.204.50.50]: FAILED! => {
    "changed": false, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_ADXrQF/ansible_module_docker_container.py\", line 1963, in <module>\n    main()\n  File \"/tmp/ansible_ADXrQF/ansible_module_docker_container.py\", line 1956, in main\n    cm = ContainerManager(client)\n  File \"/tmp/ansible_ADXrQF/ansible_module_docker_container.py\", line 1609, in __init__\n    self.present(state)\n  File \"/tmp/ansible_ADXrQF/ansible_module_docker_container.py\", line 1634, in present\n    different, differences = container.has_different_configuration(image)\n  File \"/tmp/ansible_ADXrQF/ansible_module_docker_container.py\", line 1169, in has_different_configuration\n    self.parameters.expected_exposed = self._get_expected_exposed(image)\n  File \"/tmp/ansible_ADXrQF/ansible_module_docker_container.py\", line 1555, in _get_expected_exposed\n    image_ports = [re.sub(r'/.+$', '', p) for p in (image['ContainerConfig'].get('ExposedPorts') or {}).keys()]\nTypeError: tuple indices must be integers, not str\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE", 
    "rc": 1
}
```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/cloud/docker/docker_container.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/docker/docker_container.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @chouseknecht @cove @dusdanig @joshuaconner @kassiansun @softzilla @zfil
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by kassiansun** [comment]
ansible 2.6 changed the detection logic of `ports`, could you upgrade ansible to 2.6 and try it again?

**Comment by zevarito** [comment]
I've tried `devel` with the same result.

I am not familiar with Ansible internals, from where comes
 /tmp/ansible_It7k54/ansible_module_docker_container.py?
I've noticed that ansible/modules/cloud/docker_container.py has a different
version of  `_get_expected_exposed` than the one in
/tmp/ansible_It7k54/ansible_module_docker_container.py, there is a sort of
cache in the middle?

Testing with the `pip` package I got a different backtrace for the same
module:

```
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_It7k54/ansible_module_docker_container.py", line 2081,
in <module>
    main()
  File "/tmp/ansible_It7k54/ansible_module_docker_container.py", line 2076,
in main
    cm = ContainerManager(client)
  File "/tmp/ansible_It7k54/ansible_module_docker_container.py", line 1703,
in __init__
    self.present(state)
  File "/tmp/ansible_It7k54/ansible_module_docker_container.py", line 1741,
in present
    new_container = self.container_create(self.parameters.image,
self.parameters.create_parameters)
  File "/tmp/ansible_It7k54/ansible_module_docker_container.py", line 825,
in create_parameters
    host_config=self._host_config(),
  File "/tmp/ansible_It7k54/ansible_module_docker_container.py", line 931,
in _host_config
    return self.client.create_host_config(**params)
  File "/usr/local/lib/python2.7/dist-packages/docker/api/container.py",
line 142, in create_host_config
    return utils.create_host_config(*args, **kwargs)
TypeError: create_host_config() got an unexpected keyword argument 'init'
```


2018-07-03 22:27 GMT-03:00 Kassian Sun <notifications@github.com>:
>
> ansible 2.6 changed the detection logic of ports, could you upgrade
ansible to 2.6 and try it again?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.




--
Alvaro

**Comment by kassiansun** [comment]
Oh, there's a blocking #40839 on `init` option compatibility issue. Can you upgrade `docker-py` on remote machine, or patch `ansible` manually per #40839 ?

**Comment by zevarito** [comment]
@kassiansun  which version exactly should I upgrade `docker-py`?

I've upgraded to `1.10.6` which is the latest available on my remote host through `pip` and it worked, but as soon as I checkout Ansible `v2.6.0rc5` branch instead of `devel` and tried again it didn't work, so I went back with `devel` and tried again from there but it doesn't worked anymore and fail with the  same error.

Should I close this issue and keep posting on #40839?

Thanks.
